### ISSUE TITLE
JakartaEE 9: Jakarta Enterprise Beans 4.0 Support

### ISSUE THREAD

**Issue by tkburroughs** [OpenLiberty/open-liberty#11450]
### Support Jakarta EE 9 : Jakarta Enterprise Beans (EJB) 4.0

**Feature Overview (UFO)** : https://ibm.box.com/s/pwkrthkylw25pfmutyhgtb31rr2k9at2

#### The following will be required to support `Jakarta Enterprise Beans`

 - [x] Provide the release candidate API (`jakarta.ejb`)
 - [x] Provide ejbLite-4.0 implementation (depends on ManagedBeans:  https://github.com/openliberty/open-liberty/issues/11850)
 - [x] Deliver ejbLite-4.0 in beta / webProfile
 - [x] Provide ejbHome-4.0 implementation
 - [x] Provide ejbPersistentTimer-4.0 implementation
 - [x] Provide ejbRemote-2.0 implementation
 - [x] Provide ejbRemoteClient-4.0 implementation
 - [x] Provide mdb-4.0 implementation
 - [x] Provide ejb-4.0 convenience feature
 - [x] Enterprise Beans 4.0 always RMIC Compatible #16237
 - [x] Provide EJB Container API for jakarta packaging #18526
 - [x] Reduce requirement of PortableRemoteObject.narrow() #16238
 - [x] Jakarta EE 9 support in Yoko ORB #17742 
 - [x] Provide testing  #13717
 - [x] Deliver final API
 - [x] TCK Passing

##
#### Before proceeding to any items below (active development), this feature must be prioritized on the backlog, and have either the "Design Approved" or "No Design Approved" labels.  Follow the Feature and UFO Approval Process.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: https://ibm.box.com/s/pwkrthkylw25pfmutyhgtb31rr2k9at2
- FTS: https://github.com/OpenLiberty/open-liberty/issues/15903
- Beta Blog Post (if applicable):
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/19395

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or **N/A**.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point) **No additional SVT is required for this feature beyond what will be done by the parent feature, #9876**
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality. Approved for beta releases; covered by #9876 

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [x] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [x] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update). **Covered by Covered by #9876**, with specifics detailed here : https://github.com/OpenLiberty/open-liberty/issues/19395

## **Other deliverbles**
- [x] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or **N/A**. No guides for EJB.
- [x] **WDT** - (Leonard Theivendra). WDT work complete or N/A. Covered by #9876 ; See #11989
- [x] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen) Covered by #9876

**Comment by jhanders34** [comment]
From the UFO socialization part 1 on Feb 26, here are the notes that I took:

**Communication**
	Documentation updates. openliberty.io docs to talk about what APIS are removed, what's new, etc

**missedPersistenceTimerAction**
	Add information about the values ALL vs ONCE
	Clarify the default behavior before and after?  Is that what Emily said?  I think you recorded something as well for this, but I had a hard time understanding what needed to be updated.

**Feature Design rmicCompatible**
	Just don't do it at all. Don't need it any more.  Just do true by default with no way to change it for 4.0.

**Feature Design missedPersistenceTimerAction** 
	Concerns about changing the default behavior since customers depending on the old default would not know they need to configure something to enable it.

	This was the last slide we talked about and will be where we start for the Part 2 of the UFO socialization.

**Comment by jhanders34** [comment]
Notes from part 2 of the UFO socialization from today:

**Configuration**

- Content for these slides go away since not making changes

**Developer Experience**

- Remove the second bullet

**Deprecation & Stabilization** 

- Remove the ejbHome and ejbRemote stabilization

**Beta** 

- Remove the rmicCompatible bullet

**System test** 

- System test may need to update applications to not use APIs that have been removed.

**Migration impact** 

- Update to include jakarta.ejb search as well.  
- Add in the SessionContext.getMessageContext() method.  Doesn't look like we scan for it today.

**Comment by tkburroughs** [comment]
All updates requested from the UFO Design Review have been made and new UFO has been uploaded.

**Comment by skasund** [comment]
For got to add this earlier, but L2 has requested STE slides for this feature.  The STE template can be found at the links below.  You can use either one to create the education.  

Slide Template:  https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Github Template:  https://pages.github.ibm.com/WASL3/site/STE/about

Please upload the completed slides to the same 'STE Archive' BOX folder and let me know they're complete or provide me the Github link.   Thanks!

**Comment by tkburroughs** [comment]
@skasund The STE slides have been uploaded to BOX : https://ibm.box.com/s/uus99y5b29y0snil4hbq3s9r2u0jlv2b
Title : STE-11450-Jakarta-Enterprise-Beans-4.ppt

**Comment by skasund** [comment]
@tkburroughs   Thanks for the slides.  I've approved this feature.

**Comment by tkburroughs** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?
```
The UFO indicates that there are no changes that effect serviceability. The new enterpriseBeans-4.0 features
provide the exact same functionality as the prior ejb-3.2 features.  There are no new problems to document. 
The enterpriseBeans-4.0 implementation is the same as the prior ejb-3.2 implementation.
```

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
```
The UFO indicates that there are no changes that effect serviceability. The new enterpriseBeans-4.0 features
provide the exact same functionality as the prior ejb-3.2 features.  There are no new problems to demo.
```

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  
```
a) Brian Hanczaryk  & Jagadeeswar Gangaraju
b) No serviceability issues to evaluate; the same SVT applications as used for Jakarta EE 8 were transformed
for Jakarta EE 9; no differences in behavior related to EJB.
```

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
```
This feature has the same L2/L3 teams supporting it as the Jakarta EE 8 features.
```
5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
```
The same metrics that exist for Jakarta EE 8 are relevant for Jakarta EE 9 features as well.
There are no new ones that are added.
```

**Comment by chirp1** [comment]
Approving this epic for the doc issue at https://github.ibm.com/websphere/liberty-docs/issues/2475 See my additional comments in the doc issue.

**Comment by tkburroughs** [comment]
All work complete; closing.
